cupyx/scipy/special/_logsoftmax.py

```diff
@@ -0,0 +1,64 @@
+import cupy as cp
+
+
+_log_softmax_kernel = cp._core.ReductionKernel(
+    'T x1',
+    'T y',
+    'exp(x1)',
+    'a + b',
+    'y = log(a)',
+    '0',
+    name='log_softmax'
+)
+
+
+def log_softmax(x, axis=None):
+    """Compute logarithm of softmax function
+
+    Parameters
+    ----------
+    x : array-like
+        Input array
+    axis : int or tuple of ints, optional
+        Axis to compute values along. Default is None and softmax
+        will be  computed over the entire array `x`
+
+    Returns
+    -------
+    s : cupy.ndarry
+        An array with the same shape as `x`. Exponential of the
+        result will sum to 1 along the specified axis. If `x` is a
+        scalar, a scalar is returned
+
+    """
+
+    x_max = cp.amax(x, axis=axis, keepdims=True)
+
+    if x_max.ndim > 0:
+        x_max[~cp.isfinite(x_max)] = 0
+    elif not cp.isfinite(x_max):
+        x_max = 0
+
+    tmp = x - x_max
+
+    if tmp.dtype == cp.int8:
+        tmp = tmp.astype(cp.float16)
+    elif tmp.dtype == cp.int16:
+        tmp = tmp.astype(cp.float32)
+    elif tmp.dtype == cp.int32:
+        tmp = tmp.astype(cp.float64)
+    elif tmp.dtype == cp.int64:
+        tmp = tmp.astype(cp.float64)
+    elif tmp.dtype == cp.uint8:
+        tmp = tmp.astype(cp.float16)
+    elif tmp.dtype == cp.uint16:
+        tmp = tmp.astype(cp.float32)
+    elif tmp.dtype == cp.uint32:
+        tmp = tmp.astype(cp.float64)
+    elif tmp.dtype == cp.uint64:
+        tmp = tmp.astype(cp.float64)
```

Comment:
Would you try the following works?
```py
if tmp.dtype.kind == 'i':
    for out_dtype in [cp.float16, cp.float32, cp.float64]:
        if cp.can_cast(tmp, out_dtype):
            tmp = tmp.astype(out_dtype)
            break
```

Reply:
Hi! this doesn't work, we are getting same compilation error as before due to type casting.
<details>
<summary>Error Log (Click Please)</summary>

<p>

```python
raise CompileException(log, self.src, self.name, options,
cupy.cuda.compiler.CompileException: /tmp/tmpfknvsv0g/62e007c5affa579869ad99dfafae15f1f2ec15aa.cubin.cu(34): error: more than one instance of overloaded function "exp" matches the argument list:
            function "exp(double)"
__nv_nvrtc_builtin_header.h(149892): here
            function "exp(float)"
__nv_nvrtc_builtin_header.h(150793): here
            argument types are: (const T)

/tmp/tmpfknvsv0g/62e007c5affa579869ad99dfafae15f1f2ec15aa.cubin.cu(52): error: more than one instance of overloaded function "log" matches the argument list:
            function "log(double)"
__nv_nvrtc_builtin_header.h(149880): here
            function "log(float)"
__nv_nvrtc_builtin_header.h(150789): here
            argument types are: (_type_reduce)
```

</p>
</details>

Reply:
```
if tmp.dtype.kind in 'iu':
    for out_dtype in [cp.float16, cp.float32, cp.float64]:
        if cp.can_cast(tmp, out_dtype):
            tmp = tmp.astype(out_dtype)
            break
```

Reply:
Works fine. Thanks